File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/annotations/parser/filter/telegram/MessageEntityArrayFilterParser.java
package ru.ioleksiv.telegram.bot.core.controller.annotations.parser.filter.telegram;

import org.springframework.stereotype.Component;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.IntegerFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.telegram.MessageEntityArrayFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.telegram.UserFilter;
import ru.ioleksiv.telegram.bot.api.model.objects.std.MessageEntity;
import ru.ioleksiv.telegram.bot.core.controller.annotations.parser.filter.FilterParser;
import ru.ioleksiv.telegram.bot.core.controller.annotations.parser.finder.Finder;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.Validator;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.impl.TypeNameValidator;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.impl.UnionExtractValidatorList;

import java.util.List;
import java.util.Optional;

@Component
public class MessageEntityArrayFilterParser implements FilterParser<MessageEntityArrayFilter, List<MessageEntity>> {
    @Override
    public Validator<List<MessageEntity>> createChecker(MessageEntityArrayFilter annotation, Finder finder) {
        UnionExtractValidatorList<MessageEntity> unionExtractValidatorList = new UnionExtractValidatorList<>();

        MessageEntity.Type type = annotation.type();
        if (type.notAll()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getType()), new TypeNameValidator(type));
        }
        IntegerFilter length = annotation.length();
        if (length.value().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getLength()), finder.find(length));
        }
        StringFilter url = annotation.url();
        if (url.value().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getUrl()), finder.find(url));
        }
        UserFilter user = annotation.user();
        if (user.value().isActive()) {
            unionExtractValidatorList.add(in -> Optional.ofNullable(in.getUser()), finder.find(user));
        }

        return unionExtractValidatorList;
    }

    @Override
    public Class<MessageEntityArrayFilter> getAnnotationClass() {
        return MessageEntityArrayFilter.class;
    }
}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/controller/annotations/parser/filter/telegram/MessageEntityArrayFilterParser.kt
package io.github.oleksivio.tl.kbot.core.controller.annotations.parser.filter.telegram

import io.github.oleksivio.tl.kbot.core.annotations.filter.composite.MessageEntityArrayFilter
import io.github.oleksivio.tl.kbot.core.controller.annotations.parser.filter.FilterParser
import io.github.oleksivio.tl.kbot.core.controller.annotations.parser.finder.Finder
import io.github.oleksivio.tl.kbot.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.core.controller.handler.check.impl.TypeNameValidator
import io.github.oleksivio.tl.kbot.core.controller.handler.check.impl.UnionExtractValidatorList
import io.github.oleksivio.tl.kbot.server.api.objects.std.MessageEntity

class MessageEntityArrayFilterParser :
    FilterParser<MessageEntityArrayFilter, List<MessageEntity>> {

    override val parserAnnotationClass = MessageEntityArrayFilter::class

    override fun createChecker(annotation: MessageEntityArrayFilter, finder: Finder): Validator<List<MessageEntity>> {
        val unionExtractValidatorList =
            UnionExtractValidatorList<MessageEntity>()

        annotation.validator
            .map { validatorName -> finder.find(validatorName, MessageEntity::class) }
            .forEach { validator -> unionExtractValidatorList.add({ it }, validator) }

        val type = annotation.type
        if (type.isChosen) {
            unionExtractValidatorList.add(
                { it.type },
                TypeNameValidator(type)
            )
        }
        val length = annotation.length
        if (length.status.isActive) {
            unionExtractValidatorList.add({ it.length }, finder.find(length))
        }
        val url = annotation.url
        if (url.status.isActive) {
            unionExtractValidatorList.add({ it.url }, finder.find(url))
        }
        val user = annotation.user
        if (user.status.isActive) {
            unionExtractValidatorList.add({ it.user }, finder.find(user))
        }

        return unionExtractValidatorList
    }
}
